src/node/mod.rs

```diff
@@ -0,0 +1,1241 @@
+                dst,
+                Variant::UserMessage(content),
+                None,
+            ),
+        }
+    }
+
+    fn handle_bootstrapped_to(&mut self, addr: SocketAddr) {
+        match &mut self.stage {
+            Stage::Bootstrapping(stage) => stage.send_bootstrap_request(&mut self.core, addr),
+            Stage::Joining(_) | Stage::Approved(_) => {
+                // A bootstrapped node doesn't need another bootstrap connection
+                self.core.transport.disconnect(addr);
+            }
+            Stage::Terminated => {}
+        }
+    }
+
+    fn handle_bootstrap_failure(&mut self) {
+        assert!(matches!(self.stage, Stage::Bootstrapping(_)));
+
+        info!("Failed to bootstrap. Terminating.");
+        let _ = self.user_event_tx.send(Event::Terminated);
+        self.stage = Stage::Terminated;
+    }
+
+    fn handle_connected_to(&mut self, _addr: SocketAddr) {}
+
+    fn handle_connection_failure(&mut self, addr: SocketAddr) {
+        if let Stage::Approved(stage) = &mut self.stage {
+            stage.handle_connection_failure(&mut self.core, addr);
+        } else {
+            trace!("ConnectionFailure from {}", addr);
+        }
+    }
+
+    fn handle_new_message(&mut self, sender: SocketAddr, bytes: Bytes) {
+        let msg = match MessageWithBytes::partial_from_bytes(bytes) {
+            Ok(msg) => msg,
+            Err(error) => {
+                debug!("Failed to deserialize message: {:?}", error);
+                return;
+            }
+        };
+
+        if let Err(error) = self.try_handle_message(Some(sender), msg) {
+            debug!("Failed to handle message: {:?}", error);
+        }
+    }
+
+    fn handle_unsent_message(&mut self, addr: SocketAddr, msg: Bytes, msg_token: Token) {
+        match self.core.handle_unsent_message(addr, msg, msg_token) {
+            PeerStatus::Normal => (),
+            PeerStatus::Lost => self.handle_peer_lost(addr),
+        }
+    }
+
+    fn handle_sent_message(&mut self, addr: SocketAddr, _msg: Bytes, token: Token) {
+        trace!("Successfully sent message with ID {} to {:?}", token, addr);
+        self.core.transport.target_succeeded(token, addr);
+    }
+
+    fn handle_timeout(&mut self, token: u64) {
+        if self.core.transport.handle_timeout(token) {
+            return;
+        }
+
+        match &mut self.stage {
+            Stage::Bootstrapping(stage) => stage.handle_timeout(&mut self.core, token),
+            Stage::Joining(stage) => {
+                if stage.handle_timeout(&mut self.core, token) {
+                    let network_cfg = stage.network_cfg;
+                    self.rebootstrap(network_cfg)
+                }
+            }
+            Stage::Approved(stage) => stage.handle_timeout(&mut self.core, token),
+            Stage::Terminated => {}
+        }
+    }
+
+    fn handle_peer_lost(&mut self, peer_addr: SocketAddr) {
+        if let Stage::Approved(stage) = &mut self.stage {
```

Comment:
Generally we don't consider the peer as lost if we lose a connection to it, because it can be just temporary and it can get reconnected. We use different mechanisms. For example in `Bootstrapping` we have a timeout scheduled for that. In `Approved` we use the resend attempts.